pkg/tasks/update_template_distributions.go

```diff
@@ -0,0 +1,319 @@
+package tasks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/content-services/content-sources-backend/pkg/api"
+	"github.com/content-services/content-sources-backend/pkg/config"
+	"github.com/content-services/content-sources-backend/pkg/dao"
+	"github.com/content-services/content-sources-backend/pkg/db"
+	"github.com/content-services/content-sources-backend/pkg/models"
+	"github.com/content-services/content-sources-backend/pkg/pulp_client"
+	"github.com/content-services/content-sources-backend/pkg/tasks/payloads"
+	"github.com/content-services/content-sources-backend/pkg/tasks/queue"
+	zest "github.com/content-services/zest/release/v2024"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/slices"
+)
+
+func UpdateTemplateDistributionsHandler(ctx context.Context, task *models.TaskInfo, queue *queue.Queue) error {
+	opts := payloads.UpdateTemplateDistributionsPayload{}
+	if err := json.Unmarshal(task.Payload, &opts); err != nil {
+		return fmt.Errorf("payload incorrect type for UpdateTemplateDistributions")
+	}
+
+	logger := LogForTask(task.Id.String(), task.Typename, task.RequestID)
+	ctxWithLogger := logger.WithContext(ctx)
+
+	daoReg := dao.GetDaoRegistry(db.DB)
+	domainName, err := daoReg.Domain.Fetch(task.OrgId)
+	if err != nil {
+		return err
+	}
+
+	rhDomainName, err := daoReg.Domain.Fetch(config.RedHatOrg)
+	if err != nil {
+		return err
+	}
+
+	pulpClient := pulp_client.GetPulpClientWithDomain(ctxWithLogger, domainName)
+
+	t := UpdateTemplateDistributions{
+		orgId:          task.OrgId,
+		domainName:     domainName,
+		rhDomainName:   rhDomainName,
+		repositoryUUID: task.RepositoryUUID,
+		daoReg:         daoReg,
+		pulpClient:     pulpClient,
+		task:           task,
+		payload:        &opts,
+		queue:          queue,
+		ctx:            ctx,
+		logger:         logger,
+	}
+
+	return t.Run()
+}
+
+type UpdateTemplateDistributions struct {
+	orgId          string
+	domainName     string
+	rhDomainName   string
+	repositoryUUID uuid.UUID
+	daoReg         *dao.DaoRegistry
+	pulpClient     pulp_client.PulpClient
+	payload        *payloads.UpdateTemplateDistributionsPayload
+	task           *models.TaskInfo
+	queue          *queue.Queue
+	ctx            context.Context
+	logger         *zerolog.Logger
+}
+
+func (t *UpdateTemplateDistributions) Run() error {
+	if t.payload.RepoConfigUUIDs == nil {
+		return nil
+	}
+
+	repoConfigDistributionHref := map[string]string{} // mapping to associate each repo config to a distribution href
+
+	reposAdded, reposRemoved, reposUnchanged, allRepos, err := t.daoReg.Template.GetRepoChanges(t.payload.TemplateUUID, t.payload.RepoConfigUUIDs)
+	if err != nil {
+		return err
+	}
+
+	template, err := t.daoReg.Template.Fetch(t.orgId, t.payload.TemplateUUID)
+	if err != nil {
+		return err
+	}
+
+	date := template.Date.Format(time.DateOnly)
+	l := api.ListSnapshotByDateRequest{Date: date, RepositoryUUIDS: allRepos}
```

Comment:
This was this fix for that issue you saw. The type of the `Date` field in the template create/update requests (time.Time) is different from the one in the fetchSnapshotsForDate request (string), which is why it was still working on the latter endpoint.

Reply:
I was thinking we might want to make those consistent between all the endpoints, but I went the least intrusive change for now.

Reply:
The `fetchSnapshotsForDate` endpoint expects the "yyyy-mm-dd" and doesn't seem to work correctly for other time/date formats. If the type there was time.Time, you couldn't do "yyyy-mm-dd", but it could match the time/date format returned by our API.

Reply:
I opened a bug: https://issues.redhat.com/browse/HMS-3832